synapse/handlers/sync.py

```diff
@@ -720,6 +826,162 @@ def handle_room_entries(room_entry):
+                tags=sync_result_builder.pagination_state.tags,
+            )
+            old_pagination_value = sync_result_builder.pagination_state.value
+            include_all_tags = pagination_config.tags == SYNC_PAGINATION_TAGS_INCLUDE_ALL
+        else:
+            pagination_config = None
+            old_pagination_value = 0
+            include_all_tags = False
+
+        if sync_result_builder.pagination_state:
+            missing_state = yield self._get_rooms_that_need_full_state(
+                room_ids=[r.room_id for r in room_entries],
+                sync_config=sync_config,
+                since_token=sync_result_builder.since_token,
+                pagination_state=sync_result_builder.pagination_state,
+            )
+
+            all_tags = yield self.store.get_tags_for_user(user_id)
+
+            if sync_result_builder.since_token:
+                stream_id = sync_result_builder.since_token.account_data_key
+                now_stream_id = sync_result_builder.now_token.account_data_key
+                tag_changes = yield self.store.get_room_tags_changed(
+                    user_id, stream_id, now_stream_id
+                )
+            else:
+                tag_changes = {}
+
+            if missing_state:
+                for r in room_entries:
+                    if r.room_id in missing_state:
+                        if include_all_tags:
+                            # If we're always including tagged rooms, then only
+                            # resync rooms which are newly tagged.
+                            change = tag_changes.get(r.room_id)
+                            if change == TAG_CHANGE_NEWLY_TAGGED:
+                                r.always_include = True
+                                r.would_require_resync = True
+                                r.synced = True
+                                continue
+                            elif change == TAG_CHANGE_ALL_REMOVED:
+                                r.always_include = True
+                                r.synced = False
+                                continue
+                            elif r.room_id in all_tags:
+                                r.always_include = True
+                                continue
+
+                        if r.room_id in extras.peek:
+                            since = extras.peek[r.room_id].get("since", None)
+                            if since:
+                                tok = SyncNextBatchToken.from_string(since)
+                                r.since_token = tok.stream_token
+                            else:
+                                r.always_include = True
+                                r.would_require_resync = True
+                                r.synced = False
+                        else:
+                            r.would_require_resync = True
+
+        elif pagination_config and include_all_tags:
```

Comment:
I've kind of lost track of which ``if`` statement this relates to. I wonder the block above should be a function